File: greedy/swift/452.minimum-number-of-arrows-to-burst-balloons.swift
/*
 * @lc app=leetcode id=452 lang=swift
 *
 * [452] Minimum Number of Arrows to Burst Balloons
 */

// @lc code=start
class Solution {
    func findMinArrowShots(_ points: [[Int]]) -> Int {
        var sortedBalloons = points.sorted {
            if $0[1] == $1[1] {
                return $0[0] < $1[0]
            } else {
                return $0[1] < $1[1]
            }
        }

        print(sortedBalloons)

        var end = sortedBalloons[0][1]
        var arrows = 1

        for point in sortedBalloons[1...] {
            if point[0] <= end {
                continue
            } else {
                arrows += 1
                end = point[1]
            }
        }
        return arrows
    }
}
// @lc code=end

File: greedy/cpp/452.minimum-number-of-arrows-to-burst-balloons.cpp
/*
 * @lc app=leetcode id=452 lang=cpp
 *
 * [452] Minimum Number of Arrows to Burst Balloons
 */

// @lc code=start

#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int findMinArrowShots(vector<vector<int>>& points) {
        sort(points.begin(), points.end(), [](const vector<int> &a, const vector<int> &b) {
            return a[1] < b[1];
        });

        int prev_end = points[0][1];
        int arrowCount = 1;

        for (int i = 1; i < points.size(); ++i) {
            if (points[i][0] > prev_end) {
                arrowCount += 1;
                prev_end = points[i][1];
            }
        }

        return arrowCount;

    }
};
// @lc code=end

